### PR TITLE
[Enhancement] optimize join reorder greedy proformance

### PR DESCRIPTION
**Pull request by Seaven**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

34 table inner join, before:
```
    1191ms|-- Parse[1] 179ms
    1372ms|-- Anazlye[1] 91ms
    1479ms|-- Transform[1] 157ms
    1672ms|-- Optimizer[1] 4949ms
    1803ms|    -- Optimizer.RuleBaseOptimize[1] 110ms
    1920ms|    -- Optimizer.CostBaseOptimize[1] 4646ms
    1958ms|        -- JoinReorderLeftDeep[1] 38ms
    2017ms|        -- JoinReorderGreedy[1] 4346ms
    2029ms|            -- LeftDeep[33] 4163ms
    2111ms|            -- Bushy[33] 170ms
    6568ms|    -- Optimizer.PhysicalRewrite[1] 53ms
    6622ms|-- PlanBuilder[1] 235ms
```

after:
```
    1118ms|-- Parse[1] 250ms
    1370ms|-- Anazlye[1] 86ms
    1473ms|-- Transform[1] 125ms
    1629ms|-- Optimizer[1] 1953ms
    1705ms|    -- Optimizer.RuleBaseOptimize[1] 86ms
    1798ms|    -- Optimizer.CostBaseOptimize[1] 1733ms
    1832ms|        -- JoinReorderLeftDeep[1] 26ms
    1869ms|        -- JoinReorderGreedy[1] 1468ms
    1877ms|            -- LeftDeep[33] 1284ms
    1879ms|                -- JoinStatistics[19107] 584ms
    1955ms|            -- Bushy[33] 175ms
    3534ms|    -- Optimizer.PhysicalRewrite[1] 48ms
    3583ms|-- PlanBuilder[1] 237ms
```
core modify：
fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinOrder.java

only improve proformance， none logical update

the SQL case is:
34 count distinct, starrocks will transform to CTE + Join, and I will add special HINT to reslove special Join, to avoid extra JoinReorder

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 52 / 53 (98.11%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/JoinOrder.java|20|21|95.24%|[330]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/Statistics.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/PredicateStatisticsCalculator.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/ReorderJoinRule.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/InsertPlanner.java|11|11|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinOrder.java [hunk 1]
```diff
@@ -294,39 +293,47 @@ protected void computeCost(ExpressionInfo exprInfo) {
...
+            }
+
+            for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : expressionMap.entrySet()) {
+                if (!useColumns.contains(entry.getKey())) {
+                    continue;
+                }
+
+                if (entry.getValue().isConstantRef()) {
+                    // constant always on left
                     leftExpression.put(entry.getKey(), entry.getValue());
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinOrder.java:329]: Is there improve the performance significantly?

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinOrder.java:329]: can be twice as fast on join reorder。。。

**Review by liuyehcf — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**